Sheep v2

Baa14453 opened the pull request:
<!-- Guidelines: https://docs.spacestation14.io/en/getting-started/pr-guideline -->

![logo_sheepv2](https://github.com/user-attachments/assets/b3002b9a-34b0-44f1-834f-3f082b867ba2)

## About the PR
V2 of https://github.com/space-wizards/space-station-14/pull/30496

### Adds sheep
- Adorable fluffy mammals.
- Genetically engineered to grow fibre that can be sheared using any shearing tool (wire-cutters, barbers scissors etc.)
- Friend-shaped, unlike goats.
- Slow moving, easy to wrangle.
- Produces sheep milk.
- Sprites from TG, thank you 🙏
![sheep_preview_128](https://github.com/user-attachments/assets/a8c3d5f9-8a16-4d11-ba64-8ada4ee20ae3)

### Space Sheep
- Similar to the sheep but blue, flying, spacey.
- Grows Durathread, useful.
- New sprite I commissioned from the Space Bear artist Aleksh, thank you 🙏  
![spacesheep_preview](https://github.com/user-attachments/assets/d3838700-9081-46fc-b94a-b567db399a89)

### Rainbow Sheep
- Gay.
- Grows rainbow cannabis, will invoke space drugs if a moth eats it.
- Glowy. (Tell it your secrets)

https://github.com/user-attachments/assets/271a7dec-c00f-4cef-99c5-8d52c0b25f2b


### Sheep Plushies
Sprites also commissioned from Aleksh
Moths can eat them, please do not.
CottonBurger mechanics included.
![image](https://github.com/user-attachments/assets/aac8939d-a93d-4722-b8a0-f92c32527995)

### Shearable Component
While I call it shearing, this component is modular and basically produces a product, consuming a solution, extracted using a tool.

For a sheep, the target solution is `wool` which is added by the pre-existing `wooly` component, wooly makes sheep synthesise fiber from nutrition.
Cotton is used for the shearing product because wool doesn't exist (chicken and egg problem).
2.5u of fiber/wool produces 1 piece of cotton, 25u wool into 10 cotton.
The tool is defined using the `toolQuality`, a pre-existing component applied to tools. Sheep use `shearing`. which is a new quality applied to WireCutters, Cowcutters, plant clippers, and barber's scissors.

#### Shearable Layer.
Also included is a special sprite layer definition that lets you define a toggleable layer that has specific mob state states. I know there are some similar yamlable components existing but try as I might, I could not find a working combination in yaml only and had to add something specific.

### Spawners
- Sheep Spawner
- Rainbow Sheep Spawner
- Space Sheep Spawner
- Wooly Animal Spawner
The point of this is to easily replace goat spawners with wooly animal spawners, seamlessly fitting in sheep where goats are. Despite the illusion of balance, sheep are definitely winning.
```yml
- type: entity
  parent: MarkerBase
  id: SpawnWoolyAnimalGeneric
  name: Wooly Animal Spawner
  components:
  - type: Sprite
    layers:
    - state: green
    - state: sheep_bare
      sprite: Mobs/Animals/sheep.rsi
    - state: sheep_wool
      sprite: Mobs/Animals/sheep.rsi
  - type: EntityTableSpawner
    table: !type:GroupSelector
      children:
      - !type:GroupSelector
        weight: 80
        children:
        - id: SpawnMobSheep
        - id: SpawnMobGoat
      - !type:GroupSelector
        weight: 20
        children:
        - id: SpawnMobSheepRainbow
        - id: SpawnMobSheepSpace
```

### Crates
```yml
- type: entity
  parent: CrateLivestock
  id: CrateNPCSheep
  name: sheep crate
  description: A crate containing a single sheep.
  components:
  - type: EntityTableContainerFill
    containers:
      entity_storage: !type:GroupSelector
        children:
        - id: MobSheep
        - id: MobSheepRainbow
          weight: 0.1
        - id: MobSheepSpace
          weight: 0.1
        - id: PlushieSheep # teehee get plushied
          weight: 0.001
        - id: PlushieSpaceSheep
          weight: 0.001
```

## Why / Balance
The Sheep adds a vital missing quadrupedal mammal to the farming repertoire. It offers an alternative way of acquiring essential moth ingredients even in times of low-botanist activity, and an additional method for periods of high moth activity. Efficiently converting waste and spare vegetation into useful ingredients.

It also produces sheep milk, which we'll need once the cheese connoisseurs find us and implement every single type of cheese.

The shearable component offers a dynamic way to consume internal reagents to produce products harvested using a specified tool, while also toggling a sprite layer, this might sound custom grafted exactly for sheep, and it is, but you could also use it to for example, skinning a human alive using a specific cult weapon. Consuming blood, to produce uhh... human hide, and hiding the skin as a sprite layer...

Space Sheep seem like a natural evolution and are a rare but convenient way to get Durathread.

Rainbow Sheep implement the required rainbow representation that I'm contractually obliged to consider, also it plays into my genetically modified lore.

## Technical details
### Shearable Component
https://github.com/space-wizards/space-station-14/blob/7c9dffb4f90a8529446638ae92387cf5600d445d/Resources/Prototypes/Entities/Mobs/NPCs/animals.yml#L1174-L1180

Adding a shearable component to a _thing_ will enable you to spawn a number of resources while consuming reagent within the animal.
- targetSolutionName - This is the name of a solution already within the animal.
- shearedProductID - This is the ID of a specific stackable product that will be spawned when the animal is sheared.
- solutionPerProduct - This is how many reagents it takes to spawn 1 product.
- MaximumProductsSpawned - Specifies the maximum number of items to spawn at once, requiring you to shear the animal several times to get all the resources.
- toolQuality - This is a tool quality that is needed to shear this animal. Tools have a quality component, look up the tool you want, check what qualities applies to it.
- ShearableMarkupText - An optional LOC string that is appended to the entity's description if they can be sheared.
- UnShearableMarkupText - An optional LOC string that is appended to the entity's description if they can NOT be sheared.

### Shearable Layers
Optionally, mobs with the ShearableComponent can flag a layer to be removed when the mob cannot be sheared anymore, here's an example from the sheep:
```yml
  - type: Sprite
    drawdepth: Mobs
    layers:
    - state: sheep_bare
      map: [ "enum.DamageStateVisualLayers.Base" ]
      visible: true
      sprite: Mobs/Animals/sheep.rsi
    # this wooly layer relates to the shearable component.
    - state: sheep_wool
      map: [ "wooly" ]
      visible: true
      sprite: Mobs/Animals/sheep.rsi
  - type: GenericVisualizer
    visuals:
      enum.ShearableVisuals.Shearable:
        wooly:
          True: {visible: true}
          False: {visible: false}
      enum.MobStateVisuals.State:
        wooly:
          Alive: {state: sheep_wool}
          Dead: {state: sheep_wool_dead}
          Critical: {state: sheep_wool_dead}
```
Here `enum.ShearableVisuals.Shearable` is specifying the True and False effect of shearing a mob. When it is shearable, the `wooly` layer has it's visibility set to `true`, when the mob is not `shearable`, it's set to `false`. Shearability is calculated based on whether the remaining solution has enough volume to spawn 1 product.

## Media

### SpaceSheep sprite by Aleksh (commissioned by Baa)
![spacesheep_preview](https://github.com/user-attachments/assets/d3838700-9081-46fc-b94a-b567db399a89)

### Sheep sprite from TG Station, drawn by (presumably) [MMMiracles](https://github.com/tgstation/tgstation/pull/69318)
It's from TG so licensed as [Creative Commons 3.0 BY-SA license](https://creativecommons.org/licenses/by-sa/3.0/)
![sheep_preview_128](https://github.com/user-attachments/assets/08211c3a-39fd-41c0-9492-34e3a2a8d82f)

### Rainbow Sheep
https://github.com/user-attachments/assets/271a7dec-c00f-4cef-99c5-8d52c0b25f2b

### Sheep Plushies
I commissioned from Aleksh :heart: 
![image](https://github.com/user-attachments/assets/aac8939d-a93d-4722-b8a0-f92c32527995)

### Shearing demonstration
https://github.com/user-attachments/assets/1e22bab2-206d-4d9b-80be-d8f2097eb003

### Re-growing wool
https://github.com/user-attachments/assets/6d049404-c7ae-45f9-982f-1b721e8f7b36

### Space Sheep (They don't sparkle at the moment)
https://github.com/user-attachments/assets/3625bf26-2936-4533-9237-056a67195437

Modular description
![image](https://github.com/user-attachments/assets/58c48e5f-210e-44c8-b676-65b529729919)

Rainbow Sheep and Moths

https://github.com/user-attachments/assets/b934ff09-e517-4282-a2be-921e2f257c91



![Posterv2](https://github.com/user-attachments/assets/6d3372ba-4630-4365-bf23-d226e9ea4992)

**This is a Hand Maid PR, no AI used**
<img width="150" height="20" alt="BaaBlogBanner" src="https://github.com/user-attachments/assets/621c79d7-2841-42d4-85b1-6532ef162957" />

## Requirements
<!-- Confirm the following by placing an X in the brackets [X]: -->
- [X] I have read and am following the [Pull Request and Changelog Guidelines](https://docs.spacestation14.com/en/general-development/codebase-info/pull-request-guidelines.html).
- [X] I have added media to this PR or it does not require an ingame showcase.
<!-- You should understand that not following the above may get your PR closed at maintainer’s discretion -->

**Changelog**
<!-- Add a Changelog entry to make players aware of new features or changes that could affect gameplay.
Make sure to read the guidelines and take this Changelog template out of the comment block in order for it to show up.
Changelog must have a :cl: symbol, so the bot recognizes the changes and adds them to the game's changelog. -->
:cl:
- add: Added adorable Sheep, Space Sheep, and Rainbow Sheep. Shear them with wire-cutters.
- add: Added Sheep plushies.
- add: Added Sheep Milk, high in fat.

Baa14453 committed:
Adds sheep
Ported from TG.

Baa14453 committed:
Expands the wooly system to allow wooly creatures to be sheared.
Shearing creates cotton, with a ratio to match the nutritional value of the fibre reagent.
Look, there's no wool item and I don't want to add them, they're generically modified space sheep that grow cotton, what do you want from me??

Baa14453 committed:
Add shears icon

Baa14453 committed:
Switch to better icon

Baa14453 committed:
Decouple shearing system from wooly system.

Baa14453 committed:
Add cutting quality to barber scissors.
Means they can now shear animals and cut wires.

Baa14453 committed:
Add toggled wool to sheep

Baa14453 committed:
Move wooly layer handling to wooly system.
Otherwise it wouldn't account for wooly mobs that can't be sheared, which I think would be a problem.

Baa14453 committed:
Fix incorrect message when there's not enough solution to shear off, but there's more than 0.

Baa14453 committed:
Add shearing on left-click.

Baa14453 committed:
Add dead wool sprite
Unused for now

Baa14453 committed:
I installed a formatter teehee

Baa14453 committed:
Disable wooly layer when the mob dies.

Baa14453 committed:
No need to check life state at two places.

Baa14453 committed:
Moving to shared for prediction or something

Baa14453 committed:
Completes transition to shared, adds some more comments.

Baa14453 committed:
Switch to SharedSolutionContainer to fix merge conflict?

Baa14453 committed:
Merge branch 'master' into sheep

Baa14453 committed:
Fix accidental paste.

Baa14453 committed:
Merge branch 'master' into sheep

Baa14453 committed:
Merge branch 'space-wizards:master' into sheep

Baa14453 committed:
Merge branch 'master' into sheep

Baa14453 committed:
Moved sheep spawner into animals.yml

Baa14453 committed:
Merge remote-tracking branch 'ss14-main-repo/master' into sheep

Baa14453 committed:
Move shearable system to Animals (plural)

Baa14453 committed:
Switch visuals layer to new ToggleableVisuals

Baa14453 committed:
Fixes serialization error.
Solution should not have been a DataField.
Also fixes a typo.

Baa14453 committed:
Re-arrange a little for clarity.

Baa14453 committed:
Resolve serialisation error, better formatting and sheep default to having 25 wool.

Baa14453 committed:
Added space sheep. Sprite by Aleksh, comissioned by Baa.

Baa14453 committed:
Make Jetpack fields modifiable.
Useful for adding jetpack effects to mobs

Baa14453 committed:
Space sheep now fly, glow and produce jetpack effects

Baa14453 committed:
Merge branch 'space-wizards:master' into sheep

Baa14453 committed:
Clarifying one comment and removing an un-used value.

Baa14453 committed:
Remove unused space sheep sprite.

Baa14453 committed:
Remove unused sprite and add space sheep dead.

Baa14453 committed:
SpaceSheep inherits from sheep instead of duplicating yaml

Baa14453 committed:
Re-arrange and rename sprites for consistency and clarity.
And also add a space sheep spawner.

Baa14453 committed:
Add an optional description addon for the shearable component.

Baa14453 committed:
Move all shearable system into shared.
No longer spawns stacks of things and doesn't require stackable objects to be defined.

Baa14453 committed:
Add rainbow sheep and sheep base.
Inheriting the parent of sheep was breaking spacesheep so now they all inherit from this sort of sub-sheep thing.

Baa14453 committed:
Sheep plushies, sprites pending.

Baa14453 committed:
Merge branch 'space-wizards:master' into sheep

Baa14453 committed:
Sheep plushies
Please don't eat them ;q;

Baa14453 committed:
Rainbow Sheep spawner

Baa14453 committed:
BAAAAA

Baa14453 committed:
Working shearable layer.

Baa14453 committed:
Remove ShearableVisualsComponent, move details into ShearableComponent

Baa14453 committed:
Update license definition.
I'm just saying, the website says it has no dashes.

Baa14453 committed:
Re-use pre-existing mobstate enum instead of re-writing it.

Baa14453 committed:
Remove unused return enum

Baa14453 committed:
Clean up ShearableSystem and fix calculation bug.

Baa14453 committed:
Trim silence from sheep SFX and add reverbed space sheep versions
Space Sheep now generate Durathread and use new shearable layer.

Baa14453 committed:
I don't like the word floofy

Baa14453 committed:
Sheep crates and wooly spawner

Baa14453 committed:
Rename "sheared" to "shearable" because it didnt' make sense.

Baa14453 committed:
Merge branch 'space-wizards:master' into sheep

Baa14453 committed:
Remove extra space

Baa14453 committed:
Fix YAML errors.

Baa14453 committed:
Sheep base should be abstract

Baa14453 committed:
Remove ActiveJetpack from spacesheep (crying)

Baa14453 committed:
Set SpaceSheep price to 600

Baa14453 committed:
Revert Active Jetpack changes.

Baa14453 committed:
Fix every file ending

Baa14453 committed:
Revert changes to WoolySystem
No need to cause additional merge conflicts

Baa14453 committed:
Missed a line

Baa14453 committed:
Missed another line

beck-thompson commented on Content.Shared/Animals/ShearableComponent.cs:
```diff
@@ -0,0 +1,109 @@
+    /// </summary>
+    [DataField]
+    public string ToolQuality = string.Empty;
+
```
Remove space

beck-thompson commented on Content.Shared/Animals/ShearableComponent.cs:
```diff
@@ -0,0 +1,109 @@
+using Content.Shared.Chemistry.Components;
+using Content.Shared.Mobs;
```
Unused

beck-thompson commented on Content.Shared/Animals/ShearableComponent.cs:
```diff
@@ -0,0 +1,109 @@
+    ///     Leave undefined for no tool required.
+    /// </summary>
+    [DataField]
+    public string ToolQuality = string.Empty;
```
Use a protoid of a tool quality prototype - also make it nullable for empty

beck-thompson replied:
Yeah that would work! There is an easier way though just do:
` public ProdoId<ToolQualityPrototype>? ToolQuality;` (I also made it nullable)

But yes those are the same thing this way is just a little cleaner. All it does is verify the protoID exists and if it doesn't it yells and screams

beck-thompson commented on Content.Shared/Animals/ShearableComponent.cs:
```diff
@@ -0,0 +1,109 @@
+    ///     e.g. "She has a fleece of fluffy wool."
+    /// </summary>
+    [DataField]
+    public string ShearableMarkupText = string.Empty;
```
use a `LocId` instead - also I would change the descrition because it isn't exactly a localized string, its just a locid that gets you the localized string (do the same for the one below)

beck-thompson commented on Content.Shared/Animals/ShearableComponent.cs:
```diff
@@ -0,0 +1,109 @@
+    /// Each one represents a different reason for why the target entity cannot be sheared.
+    /// Except <c>Success</c> which means it can be sheared.
+    /// </summary>
+    public enum CheckShearReturns
```
Move the enums out of the component (leave it in the same file but just outside)

beck-thompson commented on Content.Shared/Animals/ShearableComponent.cs:
```diff
@@ -0,0 +1,109 @@
+    // A special sprite layer that changes based on the mob's life state.
+    // But can also be toggled by the shearable solution dropping blow a certain amount.
+    [ViewVariables]
+    public bool Shearable { get; set; } = false;
```
do:
```
    [DataField]
    public bool Shearable;
```

beck-thompson commented on Content.Shared/Animals/ShearableComponent.cs:
```diff
@@ -0,0 +1,109 @@
+    public bool Shearable { get; set; } = false;
+
+    [Serializable, NetSerializable]
+    public enum ShearableVisuals
```
this one as well

beck-thompson replied:
github trolled but just move it out of the component!

beck-thompson commented on Content.Shared/Shearable/ShearingDoAfterEvent.cs:
```diff
@@ -0,0 +1,11 @@
+///     Thrown whenever an animal is sheared.
+/// </summary>
+[Serializable, NetSerializable]
+public sealed partial class ShearingDoAfterEvent : SimpleDoAfterEvent { }
```
I'd probably just move this into the system file 🤔
Its fine here, but sometimes its tough to look at when there are like 1 morbillino files 😆

beck-thompson commented on Content.Shared/Animals/ShearableSystem.cs:
```diff
@@ -0,0 +1,442 @@
+using Content.Shared.Verbs;
+using Robust.Shared.Prototypes;
+using Robust.Shared.Utility;
+
```
Remove extra space

beck-thompson commented on Content.Shared/Animals/ShearableSystem.cs:
```diff
@@ -0,0 +1,442 @@
+{
+    [Dependency]
+    private readonly IPrototypeManager _prototypeManager = default!;
+
```
Remove extra

beck-thompson commented on Content.Shared/Animals/ShearableSystem.cs:
```diff
@@ -0,0 +1,442 @@
+public sealed class SharedShearableSystem : EntitySystem
+{
+    [Dependency]
+    private readonly IPrototypeManager _prototypeManager = default!;
```
Do this (follow the same pattern for the rest!):
`[Dependency] private readonly IPrototypeManager _prototypeManager = default!;`

beck-thompson commented on Content.Shared/Animals/ShearableSystem.cs:
```diff
@@ -0,0 +1,442 @@
+    /// </returns>
+    /// <seealso cref="ShearableComponent.CheckShearReturns"/>
+    public ShearableComponent.CheckShearReturns CheckShear(
+        EntityUid ent,
```
Put these all on line line

beck-thompson commented on Content.Shared/Animals/ShearableSystem.cs:
```diff
@@ -0,0 +1,442 @@
+    )
+    {
+        // If false then skip checking for a tool, otherwise return on wrong tool.
+        if (checkItem)
```
Combine the ifs into one big if to reduce nesting! If you really want to make it smaller you can do something like this:
```
if (checkItem && (usedItem == null || !_tool.HasQuality(usedItem.Value, comp.ToolQuality)))
    return ShearableComponent.CheckShearReturns.WrongTool;
        
```
You also don't need to do that casting for nullables, you can just do `.Value`

Baa14453 replied:
If statement has now changed a bit and the nesting is for two different results but let me know if I can improve it still please <3

beck-thompson commented on Content.Shared/Animals/ShearableSystem.cs:
```diff
@@ -0,0 +1,442 @@
+        }
+
+        // Test if the configured product exists.
+        if (!_prototypeManager.TryIndex(comp.ShearedProductID, out var _))
```
Replace this with `HasIndex` and if it returns false, don't bother with the `ProductError` enum just put a useful debug errror in the console and return.

beck-thompson commented on Content.Shared/Animals/ShearableSystem.cs:
```diff
@@ -0,0 +1,442 @@
+
+        // Resolves the targetSolutionName as a solution inside the shearable creature. Outputs the "solution" variable.
+        if (
+            !_solutionContainer.ResolveSolution(
```
Put this all on one line

beck-thompson commented on Content.Shared/Animals/ShearableSystem.cs:
```diff
@@ -0,0 +1,442 @@
+            return;
+
+        // Resolves the targetSolutionName as a solution inside the shearable creature. Outputs the "solution" variable.
+        if (
```
Instead of doing all this stuff over again, just call `CheckShear` again! You might also need `CheckShear` to return some outputs (use out vars) so you don't need to recheck stuff.

I'd also change `CheckShear` to just return true or false, the out variables can handle the rest like the enum

Baa14453 replied:
These are implemented, except the enum returns I am in love with enum returns

beck-thompson commented on Content.Shared/Animals/ShearableSystem.cs:
```diff
@@ -0,0 +1,442 @@
+
+        // Success message.
+        _popup.PopupClient(
+            Loc.GetString(
```
Put this all on one line

beck-thompson commented on Content.Shared/Animals/ShearableSystem.cs:
```diff
@@ -0,0 +1,442 @@
+        // Spawn product.
+        for (var i = 0; i < removedSolution.Volume.Value / productsPerSolution; i++)
+        {
+
```
Remove extra space

beck-thompson commented on Content.Shared/Animals/ShearableSystem.cs:
```diff
@@ -0,0 +1,442 @@
+    private void AddShearVerb(Entity<ShearableComponent> ent, ref GetVerbsEvent<AlternativeVerb> args)
+    {
+        if (
+            args.Using == null
```
Use `CheckShear`

beck-thompson commented on Content.Shared/Animals/ShearableSystem.cs:
```diff
@@ -0,0 +1,442 @@
+        AlternativeVerb verb =
+            new()
+            {
+                Act = () =>
```
you can do:

`                Act = () => AttemptShear(ent, user, used),`

beck-thompson commented on Content.Shared/Animals/ShearableSystem.cs:
```diff
@@ -0,0 +1,442 @@
+    ///     They indicate whether the entity can be sheared or not.
+    /// </summary>
+    /// <param name="ent">the entity containing a wooly component that will be checked.</param>
+    /// <param name="args">Arguments passed through by the ExaminedEvent.
```
You forgot to close the comment with `</param>`

beck-thompson commented on Content.Shared/Animals/ShearableSystem.cs:
```diff
@@ -0,0 +1,442 @@
+                }
+                // ALL SYSTEMS GO!
+                args.PushMarkup(
+                    Loc.GetString(
```
Make this one line

beck-thompson commented on Content.Shared/Animals/ShearableSystem.cs:
```diff
@@ -0,0 +1,442 @@
+                    break;
+                }
+                args.PushMarkup(
+                    Loc.GetString(
```
one line

beck-thompson commented on Content.Shared/Animals/ShearableSystem.cs:
```diff
@@ -0,0 +1,442 @@
+            case ShearableComponent.CheckShearReturns.InsufficientSolution:
+                // Check again if this description has been set.
+                if (string.IsNullOrEmpty(ent.Comp.UnShearableMarkupText))
+                {
```
For these, I'd get rid of the brackets but that's my personal opinion (this goes for all the one liner if statements)

Baa14453 replied:
I like the brackets they just make it a little bit clearer for me <3

beck-thompson commented on Content.Shared/Animals/ShearableSystem.cs:
```diff
@@ -0,0 +1,442 @@
+        {
+            // Remove wool layer
+            _appearance.SetData(
+                ent.Owner,
```
one line

beck-thompson commented on Content.Shared/Animals/ShearableSystem.cs:
```diff
@@ -0,0 +1,442 @@
+        {
+            // Add wool layer
+            _appearance.SetData(
+                ent.Owner,
```
one line

beck-thompson reviewed (changes requested):
Looks pretty good! Some other suggestions:

1.) Maybe make the cosmic sleep have a weirder baa sound (I don't know audio very well but make it sound like spacey)?

2.) When the sheeps drop items, make them drop in small radius and random rotation and not exactly on top of each other!

3.) The cosmic sheep seem to regrow their wool VERY quickly! Maybe tone it down a little bit?

Baa14453 committed:
Fix tool selection so null actually works for no tools.
Also uses object instead of string.
Same for shearing strings.

Baa14453 committed:
Requested formatting changes, plus some comment fixes.

Baa14453 committed:
Remove extra spaces

Baa14453 committed:
Make shearing icon customisable

Baa14453 committed:
Slightly dither spawned items for exciting fun
This implementation works flawlessly without any issues and it's even predicted.

Baa14453 committed:
No longer spawns in walls

Baa14453 committed:
Space Sheep now grow two dura-thread every 2 minutes.

Baa14453 commented on Content.Shared/Animals/ShearableSystem.cs:
```diff
@@ -0,0 +1,442 @@
+                out var solution
+            )
+        )
+            return ShearableComponent.CheckShearReturns.SolutionError;
```
I can't return without the enum, it's the return type of this function.
I could have just a generic error enum value maybe?

Baa14453 replied:
```c#
        // Resolves the targetSolutionName as a solution inside the shearable creature. Outputs the "solution" variable.
        if (!_solutionContainer.ResolveSolution(ent, comp.TargetSolutionName, ref comp.Solution, out var solution))
        {
            Log.Debug($"An entity with the shearable component specifies an invalid shearedProductID: {comp.ShearedProductID}");
            return CheckShearReturns.Error;
        }
```

Baa14453 committed:
The REST of the requested changes ❤️

Baa14453 commented:
@beck-thompson is actually ready for review this time <3 <3 <3

beck-thompson commented on Content.Shared/Animals/ShearableSystem.cs:
```diff
@@ -0,0 +1,401 @@
+    ///     Handles shearing when the player left-clicks an entity.
+    ///     Doesn't run any checks, those are handled by AttemptShear.
+    /// </summary>
+    private void OnClicked(Entity<ShearableComponent> ent, ref InteractUsingEvent args)
```
Rename this to `OnInteractUsingEvent`

beck-thompson commented on Content.Shared/Animals/ShearableSystem.cs:
```diff
@@ -0,0 +1,401 @@
+using Content.Shared.Chemistry.Components;
+using Content.Shared.DoAfter;
+using Content.Shared.Examine;
+using Content.Shared.FixedPoint;
```
Put these in alphabetical order

beck-thompson commented on Content.Shared/Animals/ShearableSystem.cs:
```diff
@@ -0,0 +1,401 @@
+        }
+
+        // Test if the configured product exists.
+        if (!_prototypeManager.TryIndex(comp.ShearedProductID, out shearedProduct))
```
This is unnecessary, the `EntProtoId` will already guarantee the proto id exists, just use `Index` instead. Also don't get this here, get it right before you spawn in the other function (if it could fail you should do it here but it can't so no need)

Baa14453 replied:
I want to keep it in CheckShear because the out var is used elsewhere, like to reference the product name in a string.

beck-thompson commented on Content.Shared/Animals/ShearableSystem.cs:
```diff
@@ -0,0 +1,401 @@
+/// </summary>
+public sealed class SharedShearableSystem : EntitySystem
+{
+    [Dependency] private readonly IPrototypeManager _prototypeManager = default!;
```
Instead of saying `_prototypeManager` or `_doAfterSystem` (this goes for all of them though) just do `_proto` and `doAfter` - the system / manager at the end isn't necessary

beck-thompson commented on Content.Shared/Animals/ShearableSystem.cs:
```diff
@@ -0,0 +1,401 @@
+        // If checkItem is true, and comp.ToolQuality has been set then check return on wrong tool.
+        if (checkItem && comp.ToolQuality is not null)
+        {
+            if (// Is the player holding an item?
```
Combine these into one like this:

```
        // If checkItem is true, and comp.ToolQuality has been set then check return on wrong tool.
        if (checkItem && comp.ToolQuality != null &&
            // comment here if you want it
            (usedItem == null || !_tool.HasQuality(usedItem.Value, comp.ToolQuality)))
            return CheckShearReturns.WrongTool;
```

beck-thompson commented on Content.Shared/Animals/ShearableSystem.cs:
```diff
@@ -0,0 +1,401 @@
+        // Resolves the targetSolutionName as a solution inside the shearable creature. Outputs the "solution" variable.
+        if (!_solutionContainer.ResolveSolution(ent, comp.TargetSolutionName, ref comp.Solution, out shearingSolution))
+        {
+            Log.Debug($"An entity with the shearable component specifies an invalid shearedProductID: {comp.ShearedProductID}");
```
Looks like you don't actually need to log anything here, the `ResolveSolution` will error out by itself if there is an issue

beck-thompson commented on Content.Shared/Animals/ShearableSystem.cs:
```diff
@@ -0,0 +1,401 @@
+            case CheckShearReturns.Success:
+                // ALL SYSTEMS GO!
+                break;
+            case CheckShearReturns.WrongTool:
```
If your going to use this enum you might as well have a custom popup response for the wrong tool

Baa14453 replied:
I'm unsure, it appears when feeding them ;q;
<img width="736" height="854" alt="image" src="https://github.com/user-attachments/assets/1a4bff2e-9a0d-4ad4-80b2-2d3f90803ad3" />

Baa14453 replied:
1. Pizzas contain everything a sheep needs for a well-rounded balanced diet
2. That's a baby sheep, I don't know where they come from and no one will tell me
3. I'll take it out for now but i'll commit this if you want to see it

beck-thompson commented on Content.Shared/Animals/ShearableSystem.cs:
```diff
@@ -0,0 +1,401 @@
+        }
+
+        // Build arguments for calling TryStartDoAfter
+        var doargs = new DoAfterArgs(EntityManager, userUid, 5, new ShearingDoAfterEvent(), ent, ent, used: toolUsed)
```
`doArgs`

beck-thompson commented on Content.Shared/Animals/ShearableSystem.cs:
```diff
@@ -0,0 +1,401 @@
+    /// </summary>
+    private void OnSheared(Entity<ShearableComponent> ent, ref ShearingDoAfterEvent args)
+    {
+
```
Remove extra space

beck-thompson commented on Content.Shared/Animals/ShearableSystem.cs:
```diff
@@ -0,0 +1,401 @@
+    /// </summary>
+    private void AddShearVerb(Entity<ShearableComponent> ent, ref GetVerbsEvent<AlternativeVerb> args)
+    {
+
```
Remove extra space

beck-thompson commented on Content.Shared/Animals/ShearableSystem.cs:
```diff
@@ -0,0 +1,401 @@
+                // Default to empty string, if we just can't resolve the tool quality for whatever reason localisation have a blank variable..
+                var toolQuality = string.Empty;
+                // If a ToolQuality has been specified set it's name to toolQuality so it appears in localisation.
+                if (_prototypeManager.TryIndex(ent.Comp.ToolQuality, out var toolQualityProto, false) && toolQualityProto is not null)
```
`&& toolQualityProto is not null` is not necessary, if tryIndex returns true then the proto is not null

beck-thompson commented on Content.Shared/Animals/ShearableSystem.cs:
```diff
@@ -0,0 +1,401 @@
+                if (_prototypeManager.TryIndex(ent.Comp.ToolQuality, out var toolQualityProto, false) && toolQualityProto is not null)
+                {
+                    // Tool quality names are a Loc string so look up that and lower-case it.
+                    toolQuality = Loc.GetString(toolQualityProto.Name).ToLower();
```
this stuff is sussy, I'm not sure those will ever have Loc strings 🤔

beck-thompson commented on Content.Shared/Animals/ShearableSystem.cs:
```diff
@@ -0,0 +1,401 @@
+                    // Tool quality names are a Loc string so look up that and lower-case it.
+                    toolQuality = Loc.GetString(toolQualityProto.Name).ToLower();
+                    // If a Loc string isn't found then it will just return the same ID, which means it hasn't been configured right so just set it empty and don't use it.
+                    if (string.Equals(toolQuality, toolQualityProto.Name.ToLower()))
```
Just error if this occurs, don't try to smooth it over be defensive and error out for this kind of stuff so people realize its an issue

beck-thompson commented on Content.Shared/Animals/ShearableSystem.cs:
```diff
@@ -0,0 +1,401 @@
+            return;
+        }
+
+        var uid = ent.Owner;
```
Unused

beck-thompson commented on Content.Shared/Animals/ShearableComponent.cs:
```diff
@@ -0,0 +1,150 @@
+    ///     This is just for caching the resolved solution.
+    /// </summary>
+    [ViewVariables(VVAccess.ReadOnly)]
+    public Entity<SolutionComponent>? Solution;
```
This is sussy 🤔

I don't have time but there has to be a better way of doing this - caching stuff is kind of risky and leads to bugs. This isn't run often so there is little benefit

beck-thompson commented on Content.Shared/Animals/ShearableComponent.cs:
```diff
@@ -0,0 +1,150 @@
+///     Lets an entity be sheared by a tool to consume a reagent to spawn an amount of an item.
+///     For example, sheep can be sheared to consume wool Solution to spawn cotton.
+/// </summary>
+
```
Remove space

Baa14453 committed:
Make the shearing verb customisable.
It's used in some strings, or might be one day.

Baa14453 committed:
Bunch of requested changes, also adds popup when wrong tool used.
But i'll probably end up removing it because it happens for other clicks, like feeding the sheep.

Baa14453 committed:
Remove weird solution entity caching and instead return it from CheckShear.

Baa14453 committed:
Requested changes.

Baa14453 committed:
Merge branch 'space-wizards:master' into sheep

Baa14453 committed:
Used proper random spawning prediction.
Thanks to #39661

beck-thompson reviewed (changes requested):
Can you also rename wooly component (and make sure to add this to breaking changes) to something more generic like produce reagent from hunger or somethin I duno 😆

Baa14453 committed:
Requested changes
Fixed param name shearingSolutionEnt.
Removed unused popup line, leave comment.
Verb is now always added but disabled when appropriate. TODO look into weird butcher verb.

Baa14453 commented on Content.Shared/Animals/ShearableSystem.cs:
```diff
@@ -0,0 +1,392 @@
+            return;
+
+        // If we're not using a tool then Using will be null so we need to check it quickly.
+        EntityUid? toolUsed = null;
```
Even if fed into CheckShear `args.using` might still be null on the other side, and I need to feed it's value into `verb`.
I tried it like this:
<img width="1369" height="994" alt="image" src="https://github.com/user-attachments/assets/af2921cf-7e10-461e-b914-03b71c7325b3" />


Depending on the configuration, you don't necessarily need to be holding anything to shear something, so `args.using` being null should be expected.
If you have any ideas let me know

Baa14453 replied:
nvmd consulted some experts (friends in a discord) and I think I got it :)
I just left out .value and didn't need the check.

Baa14453 committed:
Re-arrange AddShearVerb a bit to remove a redundant null check.

Baa14453 committed:
Remove uneeded check.

Baa14453 committed:
Switch from TryIndex to Index when checking ToolQuality

Baa14453 commented on Content.Shared/Animals/ShearableComponent.cs:
```diff
@@ -0,0 +1,151 @@
+///     Each one represents a different reason for why the target entity cannot be sheared.
+///     Except <c>Success</c> which means it can be sheared.
+/// </summary>
+public enum CheckShearReturns
```
Should I just remove the `InsufficientSolution` popup? If I just return false I can't distinguish between Wrong Tool (which creates excessive popups) or insufficient solution.
If a sheep has no wool and you click it with the wrong tool, I'll just get False.
<img width="566" height="356" alt="image" src="https://github.com/user-attachments/assets/3800bb35-bd42-4ae6-866b-9bf42fafd808" />

Baa14453 committed:
Make shearable markup text nullable and null by default.

Baa14453 committed:
Switch MaximumProductsSpawned to int

Baa14453 commented:
> Can you also rename wooly component

Hesitant to touch wooly because there's a pending PR #35190 working towards merging it with udder component, one of the reasons I decoupled most of the sheep changes from wooly.

Baa14453 committed:
Merge branch 'master' into sheep

Baa14453 committed:
Revert woolyComponent changes.
I've decided it's out-of-scope and I don't want to cause mergeconflicts with any pending changes to wooly.

Baa14453 committed:
One more change removed.

TiniestShark commented:
Yeah, these are cute. 

I like the sprites, and the sheep plushie looks good and has a solid inhand. The cannabis sheep are fun too. Those all get a solid approval from me.

I'm not sure if these really fit any necessary role though, especially since moth ingredients are already painfully easy to get I would think considering cotton, cloth and fiber is in abundance on any station. They feel like a pretty fluff (ha) addition but I do appreciate the amount of thought that went into their construction. 

Also a moth eating the wool right off a sheep feels a bit barbaric but that's another talk altogether.

Princess-Cheeseballs commented:
> Also a moth eating the wool right off a sheep feels a bit barbaric but that's another talk altogether.

Already a feature even before this PR.

Baa14453 committed:
Add newline

Baa14453 committed:
Switch to FixedPoint2.Min and re-do shearing forumula documentation.

Baa14453 committed:
Make shearable spawn offset configurable.

Baa14453 committed:
Fix everything wrong with shearingSolutionToRemove
It was calculated twice.
One of the calculations just didn't work.
The comment was mostly right but I clarrified some parts and also made sure it was correct.
How confident am I? If I was advising stock trading I would call this description "Financial Advice" and then collect a large paycheck.

Baa14453 commented on Content.Shared/Animals/ShearableSystem.cs:
```diff
@@ -0,0 +1,387 @@
+        var solutionToRemove = FixedPoint2.New(
+            Math.Min(
+                (targetSolutionQuantity.Value - targetSolutionQuantity.Value % productsPerSolution) / 100,
+                maxProductsToSpawn * productsPerSolution / 100
```
Thanks for your help, I've nuked this section and moved it into CheckShear with your changes, also fixed the comment and some other stuff :heart:

beck-thompson reviewed (approved):
Approved sans Princesses stuff and a few minor things 🥳

Baa14453 committed:
Merge branch 'master' into sheep

Baa14453 committed:
Add overload functions for ChearShear and switch to propert ent<T> syntax

Baa14453 committed:
Simplify UpdateShearingLayer calculation
Clarity suggestion.

Baa14453 committed:
Add sheep plushies to fun places.

Baa14453 committed:
Fix missing line end

Baa14453 commented on Content.Shared/Animals/ShearableSystem.cs:
```diff
@@ -0,0 +1,365 @@
+                // NO WOOL LEFT.
+                _popup.PopupClient(Loc.GetString("shearable-system-no-product", ("target", Identity.Entity(ent.Owner, EntityManager)), ("product", shearedProduct.Name)), ent.Owner, userUid);
+                return;
+        }
```
Been experimenting with this today, this covers me for the popup but not for the examine scenario.
When a shearable creature is clicked, CheckShear is ran, if it has no solution, a pop-up is shown.
If a shearable creature is examined, if it has no solution, PushMarkup is added to the description box. The text for this is different from the popup.

I could work around it and pursue the boolean, but I'd be adding more checks to determine the type of failure based on the new out variable. I just think it's simpler, more readable, and more dynamic to stick to the enum returns :(

https://github.com/Baa14453/space-station-14/blob/65583692f8bf12b69a26b5a3aefa53b885f4585f/Content.Shared/Animals/ShearableSystem.cs#L344

Princess-Cheeseballs commented on Content.Shared/Animals/ShearableSystem.cs:
```diff
@@ -0,0 +1,395 @@
+            case CheckShearReturns.InsufficientSolution:
+                // NO WOOL LEFT.
+                _popup.PopupClient(Loc.GetString("shearable-system-no-product", ("target", Identity.Entity(ent.Owner, EntityManager)), ("product", shearedProduct.Name)), ent.Owner, userUid);
+                return;
```
```suggestion
        if (!CheckShear(ent, out var shearedProduct, toolUsed, out var string))
        {
            if (string != null)
                _popup.PopupClient(string, ent.Owner, userUid);
            return;
        }
```

Do this and then just handle the Loc stuff inside CheckShear, rename it to CanShear for consistency.

Princess-Cheeseballs commented on Content.Shared/Animals/ShearableSystem.cs:
```diff
@@ -0,0 +1,395 @@
+            // these must all be resolved.
+            || shearingSolutionState is null
+            || shearingSolutionEnt is null
+            || shearingSolutionToRemove is null)
```
If Checkshear is a bool and has [NotNullIf(true)] attached to each of these parameters then we don't need to resolve each of these.

Princess-Cheeseballs reviewed (changes requested):
Doubling down on CheckShear -> CanShear. It should simplify a lot of the logic.

Baa14453 committed:
Convert CheckShear to return a boolean instead of an enum.

Baa14453 committed:
Add missing documentation.

SnappingOpossum commented on Resources/Prototypes/Entities/Mobs/NPCs/space.yml:
```diff
@@ -375,6 +375,92 @@
+        mask:
+          - FlyingMobMask
+        layer:
+          - FlyingMobLayer
```
Layer and mask over-indented.

Baa14453 committed:
Merge remote-tracking branch 'ss14-main-repo/master' into sheep

Baa14453 committed:
Update Food to Edible

Baa14453 committed:
Update random calculations.
Can still be replaced once PredictedRandom is in engine

Baa14453 committed:
Rename CheckShear to CanShear

Baa14453 committed:
Fix sheep being deleted when a moth eats their wool
Fix attempting to shear sheep with all objects.
Add speech verbs and sounds for sheep.

Baa14453 committed:
Use some NotNullWhen statements to a void some additional codies <3
Remove a few other redundant things.

Baa14453 committed:
Add missing emote sounds for sheep.

Princess-Cheeseballs reviewed (changes requested):
Solutions review, looks good from a glance. 
Some of these changes are gonna have to wait until Solutions refactor is merged.

SnappingOpossum commented on Resources/Prototypes/Entities/Objects/Fun/Plushies/plushies.yml:
```diff
@@ -1188,6 +1188,115 @@
+    soundHit:
+      collection: Sheep
+      params:
+        variation: 0.05
```
An anchor and alias can be used here to reduce duplication.
```suggestion
  - type: EmitSoundOnUse
    sound: &PlushieSheepSound
      collection: Sheep
      params:
        variation: 0.05
  - type: EmitSoundOnLand
    sound: *PlushieSheepSound
  - type: EmitSoundOnActivate
    sound: *PlushieSheepSound
  - type: EmitSoundOnTrigger
    sound: *PlushieSheepSound
  - type: EmitSoundOnCollide
    sound: *PlushieSheepSound
  - type: Edible
    requiresSpecialDigestion: true
    useSound: *PlushieSheepSound
  - type: MeleeWeapon
    wideAnimationRotation: 90
    soundHit: *PlushieSheepSound
```

SnappingOpossum commented on Resources/Prototypes/Entities/Objects/Specific/Service/barber.yml:
```diff
@@ -33,6 +33,9 @@
       path: "/Audio/Weapons/bladeslice.ogg"
+  - type: Tool
+    qualities:
+      - Cutting
```
Over-indented.

SnappingOpossum commented on Resources/Prototypes/SoundCollections/sheep.yml:
```diff
@@ -0,0 +1,14 @@
+  files:
+    - /Audio/Animals/spacesheep1.ogg
+    - /Audio/Animals/spacesheep2.ogg
+    - /Audio/Animals/spacesheep3.ogg
```
Indentations can be reduced.

Baa14453 committed:
Removed unsued shearingSolutionState out var.

Baa14453 commented on Resources/Prototypes/Entities/Mobs/NPCs/animals.yml:
```diff
@@ -1118,6 +1118,174 @@
+
+- type: entity
+  name: sheep
+  parent: [MobSheepBase, SimpleMobBase]
```
Do you mean above name?

Baa14453 committed:
Requested YAML adjustments

Baa14453 committed:
Adjust documentation of CanShear to reduce duplication.

Baa14453 committed:
I missed a comma and it didn't read right

Baa14453 committed:
Suggested simplification of examine function

Baa14453 committed:
New tool quality just for shearing
Add shearing to wirecutters, barber's scissors, cowcutters, and plant clippers (these can all shear sheep).

beck-thompson reviewed (approved):
Double approved :)

Princess-Cheeseballs commented:
Let me know if you need help resolving the merge conflicts.

Baa14453 committed:
Correct ShearedProductId syntax.
Removed an unused variable.

Baa14453 committed:
Rework CanShear() function to reduce number of out vars and overide varients. for simplicity.

Baa14453 committed:
Update from Master

Baa14453 committed:
Update to new solutions system <3

Commit:
Merge remote-tracking branch 'upstream/master' into sheep

Commit:
test

Princess-Cheeseballs commented on Resources/Prototypes/Entities/Mobs/NPCs/animals.yml:
```diff
@@ -1135,6 +1135,203 @@
+  categories: [ HideSpawnMenu ]
+  components:
+  - type: Solution
+    id: wool
```
ditto

Princess-Cheeseballs reviewed (changes requested):
Couple things. I can remove the string matching in a future PR so don't worry about it.

Baa14453 committed:
Fix compiler warnings with partial class and readonly dependencies.

Baa14453 committed:
Completely rework shearing algorithm so it makes sense.
The entirety of the previous calculation was borne from my misunderstanding of what FixedPoint2 is and does and how it works.
Now it's simple and actually makes sense.

Baa14453 committed:
Rebalance sheep materials to match the solutions inside the product.

Baa14453 committed:
Yaml fix

Baa14453 commented on Content.Shared/Animals/ShearableSystem.cs:
```diff
@@ -0,0 +1,376 @@
+    private void UpdateShearingLayer(Entity<ShearableComponent> ent, Solution sol)
+    {
+        // The minimum solution required to spawn one product.
+        var minimumSol = 100 / ent.Comp.ProductsPerSolution;
```
I've replaced the entire algorithm and rebalanced the fibre based on how much is inside each product <3

Tayrtahn commented on Content.Shared/Animals/ShearableSystem.cs:
```diff
@@ -0,0 +1,344 @@
+    /// <param name="usedItem">The held item that is being used to shear the target entity.</param>
+    /// <param name="checkItem">If false then skip checking for the correct shearing tool.</param>
+    /// <returns>
+    ///     A <c>bool</c>, true means the entity can be sheared, false means it cannot.
```
```suggestion
    ///     A <see langword="bool" />, true means the entity can be sheared, false means it cannot.
```

Tayrtahn commented on Content.Shared/Animals/ShearableSystem.cs:
```diff
@@ -0,0 +1,344 @@
+    ///     Listens for changes in solution, checks if it's a wooly solution, and passes it to UpdateShearingLayer.
+    ///     Depending on the result, the woolly layer may change.
+    /// </summary>
+    /// <param name="ent">the entity containing a wooly component that will be checked.</param>
```
```suggestion
    ///     Listens for changes in solution, checks if it's a shearable solution, and passes it to UpdateShearingLayer.
    ///     Depending on the result, the shearable layer may change.
    /// </summary>
    /// <param name="ent">the entity containing a shearable component that will be checked.</param>
```

Tayrtahn reviewed (commented):
Some very minor tweaks to documentation.

Baa14453 committed:
Documentation tweaks

Baa14453 committed:
Fix any item being used to shear sheep.
I don't know why that was false

Tayrtahn commented on Content.Shared/Animals/ShearableSystem.cs:
```diff
@@ -0,0 +1,341 @@
+        {
+            // Default to empty string, if we just can't resolve the tool quality for whatever reason localisation has a blank variable.
+            var toolQuality = string.Empty;
+            var toolQualityProto = _proto.Index(ent.Comp.ToolQuality);
```
Doesn't this throw an exception if `ToolQuality` is null?

Baa14453 committed:
Switch to Loc.TryGetString, fix warning message name reference, upgrade message to warning severity from debug, add earlier return for shearable checks.

Princess-Cheeseballs reviewed (changes requested):
If you can do this one thing I'll merge it right now.

Baa14453 committed:
Make TargetSolutionName a required datafield.

Princess-Cheeseballs reviewed (approved):
I am a man of my word.
